### PR TITLE
refactor: move game objects defs to typescript

### PR DESCRIPTION
**Pull request by lmssiehdev**
types generated with app.quicktype.io

### CODE REVIEW

File: shared/defs/gameObjects/bulletDefs.ts [hunk 1]
```diff
@@ -1,10 +1,30 @@
...
 
-function defineSkin(baseType, params) {
-    return util.mergeDeep({}, BaseDefs[baseType], { baseType }, params);
+function defineSkin<T extends Partial<BulletDef>>(baseType: string, params: T) {
+    return util.mergeDeep({}, BaseDefs[baseType], { baseType }, params) as BulletDef;
 }
 
-export const BaseDefs = {
+export interface BulletDef {
+    type: string
```
Comment (leia-uwu) [inline shared/defs/gameObjects/bulletDefs.ts]: type should always be `"bullet"`

File: shared/defs/gameObjects/emoteDefs.ts [hunk 2]
```diff
@@ -1,15 +1,27 @@
...
+    Food,
+    Animals,
+    Logos,
+    Other,
+    Flags,
+    Default
+}
+
+export interface EmoteDef {
+    type: string
```
Comment (leia-uwu) [inline shared/defs/gameObjects/emoteDefs.ts]: type should always be `"emote"`

File: shared/defs/gameObjects/crosshairDefs.ts [hunk 3]
```diff
@@ -1,4 +1,13 @@
-export const CrosshairDefs = {
+export interface CrosshairDef {
+    type: string
```
Comment (leia-uwu) [inline shared/defs/gameObjects/crosshairDefs.ts]: type should always be `"crosshair"`

File: shared/defs/gameObjects/healEffectDefs.ts [hunk 4]
```diff
@@ -1,4 +1,12 @@
-export const HealEffectDefs = {
+interface HealEffectDef {
+    type: string
```
Comment (leia-uwu) [inline shared/defs/gameObjects/healEffectDefs.ts]: type should always be `"heal_effect" | "boost_effect"`

File: shared/defs/gameObjects/passDefs.ts [hunk 5]
```diff
@@ -1,4 +1,13 @@
-export const PassDefs = {
+interface PassDef {
+    type: string
```
Comment (leia-uwu) [inline shared/defs/gameObjects/passDefs.ts]: type should always be `"pass"`

File: shared/defs/mapObjectsTyping.ts [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+    tint: number
+    rot?: number
+    pos?: Vec2
+    removeOnDamaged?: boolean
+    mirrorY?: boolean
+    mirrorX?: boolean
+}
+
+type MapObstacleBound = Collider;
+// interface MapObstacleBound {
```
Comment (leia-uwu) [inline shared/defs/mapObjectsTyping.ts]: remove commented out code please

**Review by leia-uwu — CHANGES_REQUESTED** [review]

File: shared/defs/mapObjectsTyping.ts [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+            wallCount: number
+            particle: string
+            particleCount: number
+            residue: string
+            sound?: string
+        }
+        collision?: AABBWithHeight[]
+    }
+    mapObjects: Array<{
+        type?: string | (() => string)
```
Comment (leia-uwu) [inline shared/defs/mapObjectsTyping.ts:227]: remove `?` from it so you can remove the ts-expect-error on `map.ts`

Comment (lmssiehdev) [reply shared/defs/mapObjectsTyping.ts:227]: Hey thanks for taking a look but this isn't ready for review yet, and now that I've pushed the client code, it won't be for a while.

#36 is ready tho.

**Pull request closed without merging**